Add first demonstrator of inline completion

fcollonval opened the pull request:
# Description

This registers a new inline completion provider triggering completion request to OpenAI.

The inline completion is plugged into the infrastructure provided by JupyterLab (since 4.1).
The completion tokens are streamed to the user to get faster user feedback.

As for all inline completer, a couple of options are available in the application:
- enable: Whether to activate or not the completion with Mito AI
- debouncerDelay: How long to wait for the last key pressed before asking the AI for a completion
- triggerKind: Whether the completion is always triggered or only through manual request

On the server side, using traitlets
- `OpenAIProvider.api_key`: OpenAI API key, it defaults to `OPENAI_API_KEY`
- `OpenAIProvider.model`: OpenAI model, it defaults to `gpt-4o-mini`

# Testing

If the `OPENAI_API_KEY` is defined and valid, inline completion should be triggered right away.

Here is a demo:


https://github.com/user-attachments/assets/63e30d21-72b3-42e6-8671-8213b5d4f92b




The front-end part is loosely tested through an integration test `mito-ai/ui-tests/tests/aiInlineCompleter.spec.ts` in which the backend replies are mocked.

# Documentation

---

Current limitations/missing pieces:
- [ ] Backend unit tests
- [ ] Validate AI model exists
- [ ] Implement non-streaming reply
- [ ] Allow configuration of the prompt in the backend
- [ ] Clean the inline completion if it starts with the end of the existing code.

Questions:
- JupyterLab default inline shortcut acceptance is not `Tab` but `Alt+End`, what to do about that?
- Should we drop the REST handler to use only the websocket (the current implementation is parallel to the existing code)?

aarondr77 commented:
I got the completions up and running on my computer. After those changes you just pushed, the only thing I had to do was: 

1. Move the `mito-ai/utils` folder to `mito_ai/utils`. I don't have permissions to push to this PR, so can't share these changes. 
2. Go into the settings editor and enable the inline completions

I'll do a user experience review tomorrow and answer the questions you posed here. Sorry for the delay!

aarondr77 commented on mito-ai/.gitignore:
A common way that I've used inline code completions is to use a comment to share my intent with the AI and then wait for it to respond. 

It looks like right now we have the following issues:
1. The inline completion waits for me to start typing on a specific line before providing a completion. So if I write a comment and then press enter, it doesn't provide a completion. 
2. If I don't press enter to start the next line, then the inline code completion continues writing in the same line as the comment. This means the code is part of the comment instead of code I can execute! 


https://github.com/user-attachments/assets/ab18d4b5-d4a8-4ac7-a697-b16d17fb19c7

fcollonval replied:
This is fixed by the clean up suggestion.

aarondr77 commented on mito-ai/.gitignore:
For me on (Mac, Safari), tab is working to accept the completion. I think this is great because it's the standard that all other IDEs are adopting.

aarondr77 commented on mito-ai/.gitignore:
It looks like the inline completion isn't doing a good job of knowing when to add a new line. I think the AI should be doing this fine, so maybe we're just trimming the AI completion incorrectly somewhere. 

<img width="880" alt="Screenshot 2024-11-20 at 11 04 16 AM" src="https://github.com/user-attachments/assets/11ac88a5-89e8-4373-a1af-4d1b494be876">

fcollonval replied:
Solved by the clean up fonction

aarondr77 commented on mito-ai/.gitignore:
Here's a weird example where after I press tab to accept a completion, another suggestion is displayed at a previous line and a new line was added when the suggestion was applied. 

When I accepted the suggestion there was not a blank line in between `df` and `abc`

https://github.com/user-attachments/assets/21c47891-f5ce-4bdf-80ad-c64bbf1fcdea

fcollonval replied:
Bug in JupyterLab identified

aarondr77 commented on mito-ai/.gitignore:
I changed these invocation settings, but I'm still getting the AI completions by default when I stop typing. 

It looks like the settings under `History provider` work, but the settings under `Mito AI provider` don't change anything.

fcollonval replied:
I cannot reproduce this one.

aarondr77 commented on mito-ai/.gitignore:
btw, if this setting is changed so that users have to invoke it, how do users invoke? 


<img width="503" alt="Screenshot 2024-11-20 at 11 15 32 AM" src="https://github.com/user-attachments/assets/bb5111db-1046-4d46-8c38-f0ccf692bb80">

fcollonval replied:
Invoke manually is possible with the shortcut 'Alt + \'

aarondr77 replied:
This is a \

aarondr77 commented on mito-ai/.gitignore:
> JupyterLab default inline shortcut acceptance is not Tab but Alt+End, what to do about that?

On my Mac, the default is `tab` right now. Is it possible to overwrite it so that is always the default. 

This is the standard I've seen in other IDE's, so I think it would be most intuitive for users.

aarondr77 commented on mito-ai/.gitignore:
> Should we drop the REST handler to use only the websocket (the current implementation is parallel to the existing code)?

I think this would make sense! This would also let us have streaming responses for the chat which would be a nice ux fix. 

You know these intricacies better than I do though. Are there restrictions on what we can do through the web socket? For example, [you can see in this PR](https://github.com/mito-ds/mito/pull/1365/files) that we now optionally use an aws lambda function to get the AI completion if the user has not set an OpenAI key and wants to use ours instead. 

What do you think is the best approach? I trust your intuition.

fcollonval replied:
No issue to do this - it was agreed to do this in a follow-up.

aarondr77 commented on mito-ai/.gitignore:
In order to get the completions set up, I needed to go into the settings and enable the inline completer. Can we turn this on by default when the user installs the `mito-ai` package? That's why they are downloading the package after + setting their OpenAI API key after all.

fcollonval replied:
The commit c4cbb29 introduces a notification in case of first installation and disabled inline completion to activate it. Let's discuss that and the other points tomorrow.


https://github.com/user-attachments/assets/68a1b940-2b08-407a-b43c-6c5c8cfe025b

aarondr77 reviewed (commented):
This is a super exciting start! Great work getting this up so fast. 

I used the tool and left a few comments about areas for improvement. I think the main buckets are: 

1. Getting better responses from the AI
2. More seamlessly incorporating the suggested code into the code cell. I ran into a few rough edges here that made it hard to use. 

All of these comments were generated on my Mac on Safari and using the following settings: 

<img width="1282" alt="Screenshot 2024-11-20 at 9 55 01 AM" src="https://github.com/user-attachments/assets/649d167c-8d58-47e7-9931-554d3959d89a">

Super excited to polish this up and use it some more. Let me know your thoughts on next steps and how you want to proceed.

fcollonval commented:
Latest update:
- Move test in `tests` folder and drop snapshot matching
- Introduce a clean-up step for the inline suggestion to remove duplication of content with what is already typed by the user
- Simplify the code to track only the active inline completion - any previous completion will be cancelled or raised an error if access later.
- We are bitten by a bug upstream in JupyterLab: https://github.com/jupyterlab/jupyterlab/issues/17011  
  I'll push a fix for it.

Bug seen:

https://github.com/user-attachments/assets/474be4da-a9a1-4823-94a1-242f2eefbc3f

- Side effect, when the AI _correct_ the existing code there is kind of a duplication like at the end of this video in which for the last suggestion `fibonnaci` is corrected as the user code contains a typo:

https://github.com/user-attachments/assets/84b8c54c-52a2-4ba9-8348-ebb49ccf438a

fcollonval commented on mito-ai/.gitignore:
I guess linked to the identified bug in JupyterLab or resolved with the new aggressive discard of any previous inline completion cache attempts.

aarondr77 reviewed (commented):
Okay great. Awesome work. I'm super excited. I left a bunch of comments, but they are mostly just asking for more documentation in some places + asking some questions for my own understanding. I'm very open to your feedback, so if you disagree with the review comments, just lmk! 

## A few requested changes

### Moving from promise chains to async/awaits with try catch statements 
I think the only big-ish refactor that would be good is to move away from promise chains in favor of async/await calls. As far as I know, async/await with try catch blocks is the approach that modern codebases are adopting and it's what the rest of our codebase uses. Let me know if you have any strong preferences on not doing this. I didn't add a comment everywhere this occurs, so figured I would just leave this comment here. We don't need to update them all right now, but as we touch this code going forward, let's move to that. 

### More comments and documentation
In particular for Jupyter-specific implementation details, it would be great if you could add some more comments to share your expertise on how/why things work the way they do. I left some specific comments below that ask for this in specific places + shared some examples of comments I've found helpful in the past. 

### Default inline completer
I think when the MitoAI inline completer is turned on, we should turn off the history completer. The history completer does not seem super useful in my usage of the tool and I think we can do a better job of giving suggestions (or deciding not to give a suggestion) than the history completer. 

## Code Understanding Check 

Here's my understanding of the codebase, let me know where I am wrong.

0. We've created a web socket that lets us send messages back and forth from the frontend/backend. Because it's a web socket, we need to take special care of tracking which responses go with which messages. 
1. We've implemented a custom `InlineCompletionProvider` called the `MitoAIInlineCompleter`. 
2. Every 250 ms we create a new a new stream message that requests OpenAI sends back a completion. In this message, we send the cell prefix (everything before the cursor) and cell suffix as additional context. 
4. Each time we get a streamed response part, we do a bit of cleaning and send the the combined response parts so they can be displayed in the UI. 
5. We continue streaming the messages until we get a final message part back that has `done = true` and then we stop the stream. 
6. When the user presses `tab` it accepts the suggestion. When they continue typing, it rejects the suggestion. But there is a bug in JupyterLab that sometimes makes it so that when you accept the suggestion, it also shows a new suggestion in the wrong spot of the code cell. 

### Merging this PR

Now that you have write access, feel free to merge this directly into dev once you address the review. No need for another review. Then, if we see that tests are failing, you should be able to open up a cloned PR and have the tests run!

fcollonval commented on mito-ai/mito_ai/__init__.py:
```diff
@@ -44,6 +38,14 @@ def _load_jupyter_server_extension(server_app):
+    handlers = [
+        (route_pattern, OpenAICompletionHandler),
+        (
+            url_path_join(base_url, "mito-ai", "inline-completion"),
```
This is independent of the package name. As far as I recall, dash is a more common separator in url than underscore.

Let me know if you prefer the underscore.

fcollonval commented:
> Moving from promise chains to async/awaits with try catch statements

I'll have a look for that. 

I agree with your statement with one exception. If you want to start an asynchronous task in the background or within a synchronous call (like in a constructor).

fcollonval commented on mito-ai/mito_ai/models.py:
```diff
@@ -0,0 +1,94 @@
+
+
+@dataclass(frozen=True)
+class InlineCompletionItem:
+    """The inline completion suggestion to be displayed on the frontend.
+
+    See JupyterLab `InlineCompletionItem` documentation for the details.
+    """
+
+    insertText: str
+    filterText: Optional[str] = None
+    isIncomplete: Optional[bool] = None
+    token: Optional[str] = None
+    error: Optional[InlineCompletionError] = None
```
It does not need to stay compatible with JupyterLab type as that constrain is only true in the frontend. But we are free to change the interface between our provider and our backend. 

> Note we cannot import that type from JupyterLab as it is a typescript type.

fcollonval replied:
To link to the code, the actual place those interfaces must be used are in:
1. MitoAIInlineCompleter.fetch; the return object must match `IInlineCompletionList<IInlineCompletionItem>`
1. MitoAIInlineCompleter._receiveStreamChunk the `emit`ted chunk must match `InlineCompletionStreamChunk`

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/index.ts:
```diff
@@ -0,0 +1,147 @@
+      name: CONFIG_SECTION,
+      serverSettings: app.serviceManager.serverSettings
+    })
+      .then(async config => {
```
This one is an optimization. An activation method can be asynchronous. But as this extension is auto-started at the application launch, it will be part of the extensions _blocking_ the start of the application. Therefore this is a good practice to carry network request as background task as done here. If from a style point of view, you prefer using `try {} catch () {}` syntax for such case. That could be rewritten using the following pattern:

```ts
const f = async () => {
    try {
        await call_network()
    } catch(r) {
        ...
    }
}
f()
```

Let me know if you want me to change that.

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/provider.ts:
```diff
@@ -0,0 +1,403 @@
+
+    this._client
+      .initialize()
+      .then(() => {
```
This could be rewritten using the trick of defining a local async function if you want to avoid this promise chain in the constructor.

fcollonval commented on tests/mitoai_ui_tests/aiInlineCompleter.spec.ts:
```diff
@@ -0,0 +1,297 @@
+    const filename = 'inline-completer.ipynb';
+    await page.notebook.createNew(filename);
+    await page.notebook.setCell(0, 'code', 'def fib');
+    await page.keyboard.press('Alt+\\');
```
This is actually the _Option_ key on the keyboard. But code-wise, it emits `Alt` key.

fcollonval commented:
Code Understanding Check

1. We've created a web socket that lets us send messages back and forth from the frontend/backend. Because it's a web socket, we need to take special care of tracking which responses go with which messages.

:+1: 

1. We've implemented a custom InlineCompletionProvider called the MitoAIInlineCompleter.

:+1: 

1. Every 250 ms we create a new a new stream message that requests OpenAI sends back a completion. In this message, we send the cell prefix (everything before the cursor) and cell suffix as additional context.

Every time the user stop typing for 250ms (the `debouncerDelay` settings), we send a message through the web socket to initiate a code completion. And indeed for now the context is bound to the active cell (something to improve).

1. Each time we get a streamed response part, we do a bit of cleaning and send the the combined response parts so they can be displayed in the UI.

The JupyterLab inline completer is expecting the all completion to be sent at each stream tick not the chunk. Therefore indeed we do the aggregation in the mito provider and apply the cleaning on the raw aggregation. The cleaning on the raw aggregation is needed because the stream chunk may contain an empty string or only 2-3 characters. So it will be hard to apply meaningful post processing on the chunk directly.

1. We continue streaming the messages until we get a final message part back that has done = true and then we stop the stream.

:+1: (or an error)

1. When the user presses tab it accepts the suggestion. When they continue typing, it rejects the suggestion. But there is a bug in JupyterLab that sometimes makes it so that when you accept the suggestion, it also shows a new suggestion in the wrong spot of the code cell.

:+1: 
The bug is that the inline completer consider the accepted completion as new content added by the user. Hence it triggers an inline completion request but with the cursor still positioned prior to the accepted suggestion.

fcollonval commented on mito-ai/mito_ai/providers.py:
```diff
@@ -0,0 +1,158 @@
+        # Use by the frontend to reconciliate the completion with the request.
+        token = self.get_token(request)
+
+        yield InlineCompletionReply(
+            list=InlineCompletionList(
+                items=[
+                    InlineCompletionItem(insertText="", isIncomplete=True, token=token)
+                ]
+            ),
+            parent_id=request.message_id,
+        )
```
:+1:

fcollonval commented on mito-ai/mito_ai/providers.py:
```diff
@@ -0,0 +1,158 @@
+                parent_id=request.message_id,
+                response=InlineCompletionItem(
+                    insertText=chunk.choices[0].delta.content or "",
+                    isIncomplete=True,
```
Actually the `insertText` here will always be a chunk of the completion as the aggregation is done in the frontend.

> I updated the `isIncomplete` to be `!done` in the frontend :wink:

fcollonval commented on mito-ai/mito_ai/providers.py:
```diff
@@ -0,0 +1,158 @@
+        return self._client
+
+    async def _check_authentication(self) -> None:
+        # # TODO implement this
+        # async for _ in client.models.list():
+        #     logging.getLogger("ServerApp").info("%s", _)
+        #     break
+        ...
```
I'll see if I have time for that - but I plan to do it when dropping the REST handler.

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/provider.ts:
```diff
@@ -0,0 +1,403 @@
+
+      let cellId = undefined;
+      let path = context.session?.path;
+      if (context.widget instanceof NotebookPanel) {
+        const activeCell = context.widget.content.activeCell;
+        if (activeCell) {
+          cellId = activeCell.model.id;
+        }
+      }
```
Good catch

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/provider.ts:
```diff
@@ -0,0 +1,403 @@
+  }
+
+  /**
+   * Extract prefix from request, accounting for context window limit.
```
For now we don't - this is something to do but I don't know if it should be done here in the frontend or rather in the backend before calling OpenAI

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/provider.ts:
```diff
@@ -0,0 +1,403 @@
+      }
+      const messageId = ++this._counter;
+
+      const stream = true;
```
I dropped it as suboptimal - good catch

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/provider.ts:
```diff
@@ -0,0 +1,403 @@
+          break;
+        }
+      }
+      if (stream && result.list.items[0]?.token) {
+        this._currentToken = result.list.items[0].token;
+        this._currentStream = new Stream<
+          MitoAIInlineCompleter,
+          InlineCompletionStreamChunk
+        >(this);
+      }
```
Correct

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/provider.ts:
```diff
@@ -0,0 +1,403 @@
+      return Promise.reject('Mito AI provider is disposed.');
+    }
+
+    await this._completionLock.promise;
+
+    this._completionLock = new PromiseDelegate<void>();
```
No we wait for the acknowledge answer to be received. That acknowledge answer is the one providing the unique `token` of the suggestion.

This is needed as the last version of the code only track the latest completion request.

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/provider.ts:
```diff
@@ -0,0 +1,403 @@
+    await this._completionLock.promise;
+
+    if (this._currentToken !== token) {
+      // New completion may be triggered before the code have started streaming
+      // the previous one. So not raising an error here.
+      console.debug(`No stream found for token '${token}'.`);
```
Yes

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/provider.ts:
```diff
@@ -0,0 +1,403 @@
+    }
+
+    for await (const chunk of this._currentStream!) {
+      if (this._currentToken !== token) {
+        break;
+      }
```
Actually it may happen. If JupyterLab has started consuming that stream and a new request comes in, the `_currentToken` will change and we stop the stream has useless has the previously completion should be discarded.

fcollonval commented on mito-ai/src/Extensions/InlineCompleter/provider.ts:
```diff
@@ -0,0 +1,403 @@
+    if (this._currentPrefix) {
+      if (
+        cleanedCompletion.startsWith(this._currentPrefix) ||
+        this._currentPrefix.startsWith(cleanedCompletion)
```
the second case is due to the streaming of the completion. Taking back the first example:

`_currentPrefix -> "pri",`

The first completion chunk may be `"pr"` and so we don't want to display to the user `"pripr"`. Then when the second chunk arrives, like "int". The full completion will be `print` and we will have a proper cleaned completion `nt`.

fcollonval replied:
> Now that I think about this, maybe a better check to do is to check if we recalculated the currentPrefix, would it match exactly to the currentPrefix when the request was made.
> 
> This would account for the case where:
> 
>     1. The user continued typing
> 
>     2. The user deleted/changed the part of the prefix.
> 
> 
> Thoughts? I might be missing something.

Currently we cannot account easily for the new code changes by the user as any changes will discard the inline suggestion. But indeed I saw that copilot does that. Something that would require changing with JupyterLab core code. But certainly worth the improved user experienced.

fcollonval commented:
@aarondr77 I think I addressed all comments. I'll keep the PR updated for another day or so. Then I'll merge and fix right after the CI (likely broken I fear).

aarondr77 commented:
Sounds great @fcollonval ! Nice Job on this big PR
